--- cocotron/Foundation/platform_posix/NSSocket_bsd.m.orig	2015-11-17 14:49:45.126666665 +0800
+++ cocotron/Foundation/platform_posix/NSSocket_bsd.m	2015-11-17 14:50:06.029999992 +0800
@@ -96,7 +96,7 @@
 - initConnectedToSocket: (NSSocket **)otherX
 {
     int pipes[2];
-    if (pipe(pipes)) {
+    if (!pipe(pipes)) {
         *otherX = [[[isa alloc] initWithDescriptor:pipes[0]] autorelease];
         return [self initWithDescriptor:pipes[1]];
     } else {
